sympy/solvers/tests/test_solveset.py

```diff
@@ -2014,6 +1915,121 @@ def test_solve_logarithm():
+
+    assert solveset_real(2*x + 5 + log(3*x - 2), x) == \
+        FiniteSet(Rational(2, 3) + LambertW(2*exp(-Rational(19, 3))/3)/2)
+
+    assert solveset_real(3*x + log(4*x), x) == \
+        FiniteSet(LambertW(Rational(3, 4))/3)
+
+    assert solveset_real(a/x + exp(x/2), x) == \
+        FiniteSet(2*LambertW(-a/2)) - FiniteSet(0)
+    assert solveset_real(x*exp(x) - 1, x) == FiniteSet(LambertW(1))
+    eq = (x*exp(x) - 3).subs(x, x*exp(x))
+    assert solveset_real(eq, x) == \
+        FiniteSet(LambertW(3*exp(-LambertW(3))))
+
+    assert solveset_real(x*log(x) + 3*x + 1, x) == \
+        Intersection(Interval(0, oo), FiniteSet(exp(-3 + LambertW(-exp(3)))))
+    # issue 4271
+    assert solveset_real((a/x + exp(x/2)).diff(x, 2), x) == Intersection(
+        Interval(0, oo), FiniteSet(6*LambertW((-a)**(S(1)/3)/3))) - FiniteSet(0)
+    assert solveset_real(x**3 - 3**x, x) == FiniteSet(
+        -3*LambertW(-log(3)/3)/log(3), -3*LambertW(-log(3)/3, -1)/log(3))
+    assert solveset_real(x**2 - 2**x, x) == solveset_real(-x**2 + 2**x, x)
+
+    assert solveset_real((a/x + exp(x/2)).diff(x), x) == \
+        Intersection(Interval(0, oo),
+            FiniteSet(4*LambertW(sqrt(2)*sqrt(a)/4))) - FiniteSet(0)
+    a = -1
+    assert solveset_real((a/x + exp(x/2)).diff(x), x) == S.EmptySet
+    a = 1
+    assert solveset_real((a/x + exp(x/2)).diff(x), x) == FiniteSet(
+        4*LambertW(sqrt(2)/4))
+    assert solveset_real(x**x - 2, x) == FiniteSet(exp(LambertW(log(2))))
+
+
+@XFAIL
+def test_other_solve_lambert_():
+    a = Symbol('a')
+
+    assert solveset_real(-a*x + 2*x*log(x), x) == \
+        Union(FiniteSet(0), Intersection(S.Reals, FiniteSet(exp(a/2))))
+    assert solveset_real(log(log(x - 3)) + log(x-3), x) == FiniteSet(
+        exp(LambertW(1)) + 3)
+
+
+def test_solve_bivariate():
+
+    assert solveset_real((x**2 - 2*x + 1).subs(x, log(x) + 3*x), x) == \
+        FiniteSet(LambertW(3*S.Exp1)/3)
+    assert solveset_real((x**2 - 2*x + 1).subs(x, (log(x) + 3*x)**2 - 1), x) == \
+        FiniteSet(LambertW(3*exp(-sqrt(2)))/3, LambertW(3*exp(sqrt(2)))/3)
+    assert solveset_real((x**2 - 2*x - 2).subs(x, log(x) + 3*x), x) == \
+        FiniteSet(LambertW(3*exp(1 + sqrt(3)))/3, LambertW(3*exp(-sqrt(3) + 1))/3)
+
+
+@XFAIL
+def test_other_solve_bivariate():
+    eq = 2*(3*x + 4)**5 - 6*7**(3*x + 9)
+    result = solveset_real(eq, x)
+    ans = FiniteSet((log(2401) +
+                     5*LambertW(-log(7**(7*3**Rational(1, 5)/5))))/(3*log(7))/-1)
+    assert result == ans
+    assert solveset_real(eq.expand(), x) == result
+
+    # coverage test
+    assert solveset_real(tanh(x + 3)*tanh(x - 3) - 1, x) == EmptySet()
+    assert solveset_real(3**cos(x) - cos(x)**3) == FiniteSet(
+        acos(-3*LambertW(-log(3)/3)/log(3)))
+
+
+@XFAIL
+def test_other_lambert():
+    a = S(6)/5
+    assert solveset_real(x**a - a**x, x) == FiniteSet(
+        a, -a*LambertW(-log(a)/a)/log(a))
```

Comment:
When `a` is greater than 1, `solve` knows that there is a solution:
```
>>> var('a',positive=True)
a
>>> solveset_real(x**(1+a) - (1+a)**x, x)
ConditionSet(x, Eq(x**(a + 1) - (a + 1)**x, 0), Reals)
>>> solve(x**(1+a) - (1+a)**x, x)
[-(a + 1)*LambertW(-log(a + 1)/(a + 1))/log(a + 1)]
```
(Symmetry gives the second solution.)

Reply:
Another thing that I see here is `_lambert()` is able to handle most of such types
 
```python
>>> a, x = symbols('a x', positive=True)
>>> _solve_lambert(x**(2) - (2)**x, x, gens)
[2, -2*LambertW(-log(2)/2, -1)/log(2)]    # ok
>>> _solve_lambert(x**(a) - (a)**x, x, gens)
[-a*LambertW(-log(a)/a)/log(a)]    # ok

>>> u = Dummy('u', positive=True)
>>> _solve_lambert(u**(a) - (a)**u, u, gens)   # NotImplementedError (interesting?)
>>> _solve_lambert(x**(1+a) - (1+a)**x, x, gens)    # NotImplementedError
```
I did not have a thorogh look at the cause but the last two errors suggets that maybe `_lambert()` is missing something. Isn't it?

Reply:
Yeah change `if mainpow and symbol in mainpow.exp.free_symbols:` to `if mainpow`. I think it doesn't matter if exp has the symbol or not

Reply:
And for adding symmetric solution you can do something like this
```
if mainpow:
            lhs = collect(lhs, mainpow)
            if lhs.is_Mul and rhs != 0:
                soln = _lambert(expand_log(log(lhs) - log(rhs)), symbol)
            elif lhs.is_Add:
                # move all but mainpow-containing term to rhs
                other = lhs.subs(mainpow, 0)
                mainterm = lhs - other
                rhs = rhs - other
                diff = log(mainterm) - log(rhs)
                soln = _lambert(expand_log(diff), symbol)
                if rhs.is_Pow:
                    if mainterm.base == rhs.exp and mainterm.exp == rhs.base:
                        soln = Union(soln, solveset(mainterm.base - mainterm.exp, symbol))
```
in _solve_lambert